scripts/Muon/GUI/Common/threading_manager.py

```diff
@@ -0,0 +1,282 @@
+
+        # Callback called when all threads have executed
+        self.thread_complete_callback = callback_on_threads_complete
+        # Callback called when a single thread updates its progress
+        self.progress_callback = callback_on_progress_update
+        # Callback called when an error is thrown from a thread
+        self.error_callback = callback_on_thread_exception
+
+        self.fn = fn
+        self.kwarg_list = kwarg_list
+
+        # maintain a list of running threads to allow for cancelling
+        self._threads = []
+
+        self._results = {key: [] for key in self.kwarg_list.keys() + ['results']}
+        self._failed_results = {key: [] for key in self.kwarg_list.keys()}
+
+        self._progress = 0.0
+
+        self._num_threads = num_threads  # total number of threads started
+        self._thread_count = 0  # current threads running
+
+        self.mutex = QtCore.QMutex()
+
+        self.verbose = verbose
+
+    def _cancel_threads(self):
+        for thread in self._threads:
+            if thread.isRunning():
+                thread.cancel()
+
+    def _clear_threads(self):
+        self._cancel_threads()
+        self._threads = []
+        self._thread_count = 0
+
+    def _clear_results(self):
+        self._progress = 0.0
+        self._results = {key: [] for key in self.kwarg_list.keys() + ['results']}
+        self._failed_results = {key: [] for key in self.kwarg_list.keys()}
+
+    def cancel(self):
+        self._clear_threads()
+        self.cancelled.emit()  # for unit tests
+
+    def clear(self):
+        self._clear_threads()
+        self._clear_results()
+
+    def is_running(self):
+        return self._thread_count > 0
+
+    def start(self):
+        if not self.is_running():
+            self.clear()
+            self._thread_count = self._num_threads
+            self.on_thread_progress(0.0)
+
+            thread_list = split_kwarg_list(self.kwarg_list, self._num_threads)
+            for i, arg in enumerate(thread_list):
+                worker = Worker(self.fn, **arg)
+                self.connect_worker_signals(worker)
+                self._threads += [worker]
+                if self.verbose:
+                    print("\tStarting thread " + str(i + 1) + " with arg : ", arg)
+                worker.start()
+        else:
+            raise RuntimeError("Cannot start threads, "
+                               "WorkerManager has active threads (call cancel() or clear() first)")
+
+    def connect_worker_signals(self, worker):
+        worker.signals.started.connect(self.on_thread_start)
+        worker.signals.result.connect(self.on_thread_result)
+        worker.signals.finished.connect(self.on_thread_complete)
+        worker.signals.progress.connect(self.on_thread_progress)
+        worker.signals.error.connect(self.on_thread_exception)
+        worker.signals.cancelled.connect(self.on_thread_cancelled)
+
+    def on_thread_start(self):
+        pass
```

Comment:
I think this and the next function could benefit from a comment that they are intentionally left empty.